gqcp/include/Mathematical/Representation/Tensor.hpp

```diff
@@ -419,23 +419,40 @@ class Tensor:
      *
      *  @return This tensor as a matrix of the specified dimensions, ordered using C-like ordering.
      */
-    const GQCP::Matrix<Scalar> reshape(const size_t rows, const size_t cols) const {
+    GQCP::Matrix<Scalar> reshape(const size_t rows, const size_t cols) const {
 
+        // The dimensions of the tensor have to match the dimensions of the target matrix.
+        if (rows * cols != this->dimension(0) * this->dimension(1) * this->dimension(2) * this->dimension(3)) {
```

Comment:
Maybe we could add an API `Tensor::numberOfElements` instead of this manual calculation?